### PR TITLE
rpc: improve error codes for internal server errors

### PR DESCRIPTION
**Pull request by niczy**
This PR changes the error codes emitted by the RPC server in certain situations:

||old error code|new error code|
|-|-|-|
|handler panic|-32000|-32603|
|result marshaling error|-32000|-32603|
|attempt to subscribe via HTTP|-32000|-32001|

### CODE REVIEW

**Comment by dbadoy** [comment]
JSON-RPC spec say, 
```
-32603 Internal error Internal JSON-RPC error.
```
Any have reason use code `-32605` ?

**Comment by dbadoy** [comment]
https://github.com/ethereum/go-ethereum/blob/4b9c307d26f19ff228cca379aeace974f26df6ce/rpc/handler.go#L356-L358

**Comment by fjl** [comment]
I think error code -32603 would be more correct for this case.

But there are too many other unrelated changes in this PR. Please remove them.

File: rpc/client_test.go [hunk 1]
```diff
@@ -84,8 +84,8 @@ func TestClientErrorData(t *testing.T) {
 	// Check code.
 	if e, ok := err.(Error); !ok {
 		t.Fatalf("client did not return rpc.Error, got %#v", e)
-	} else if e.ErrorCode() != (testError{}.ErrorCode()) {
-		t.Fatalf("wrong error code %d, want %d", e.ErrorCode(), testError{}.ErrorCode())
+	} else if e.ErrorCode() != internalServerErrorCode {
+		t.Fatalf("wrong error code %d, want %d", e.ErrorCode(), internalServerErrorCode)
 	}
 	// Check data.
 	if e, ok := err.(DataError); !ok {
```
Comment (niczy) [inline rpc/client_test.go:91]: Not sure if we intend to return internal error data from services. Didn't find any of our service implementation implement ErrorData.

**Comment by niczy** [comment]
> Wrapped the error in `runMethod` and `ErrNotificationsUnsupported` too. But didn't have anyway to test `ErrNotificationsUnsupported`, since handler always has `allowSubscribe` set to true.

**Review by JurbinM — APPROVED** [review]

**Comment by MariusVanDerWijden** [comment]
Looks to be conforming to https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md

File: rpc/client_test.go (hunk 1 above)
Comment (fjl) [reply rpc/client_test.go:91]: I'm not sure what you are asking here. This test just verifies that the error `"data"` field is returned at all when the server sends it.

File: rpc/handler.go [hunk 2]
```diff
@@ -388,7 +387,7 @@ func (h *handler) handleSubscribe(cp *callProc, msg *jsonrpcMessage) *jsonrpcMes
 func (h *handler) runMethod(ctx context.Context, msg *jsonrpcMessage, callb *callback, args []reflect.Value) *jsonrpcMessage {
 	result, err := callb.call(ctx, msg.Method, args)
 	if err != nil {
-		return msg.errorResponse(err)
+		return msg.errorResponse(&internalServerError{cause: err})
```
Comment (fjl) [inline rpc/handler.go:390]: This change is not correct. Here, `err` is the error value returned by the RPC method implementation. When the method returns an error that implements `rpc.Error` or `rpc.DataError`, we want to return the code and data contained in `err`.

However, with this change, the error code is overwritten.

File: rpc/client_test.go (hunk 1 above)
Comment (niczy) [reply rpc/client_test.go:95]: My question was whether this is by design to return the error from the internal service directly. I didn't find existing services return `rpc.Error`. Thanks for confirming :-)

File: rpc/handler.go (hunk 2 above)
Comment (niczy) [reply rpc/handler.go]: what if the RPC method returns non `rpc.Error`, should we wrap it with internalServerError?

File: rpc/handler.go [hunk 3]
```diff
@@ -390,7 +390,14 @@ func (h *handler) handleSubscribe(cp *callProc, msg *jsonrpcMessage) *jsonrpcMes
...
 	if err != nil {
-		return msg.errorResponse(err)
+		if ec, ok := err.(Error); ok {
+			return msg.errorResponse(ec)
+		} else {
+			return msg.errorResponse(&internalServerError{
+				code: errorcodeInternalServerError,
+				message: err.Error(),
+			})
+		}
```
Comment (fjl) [inline rpc/handler.go]: This change is not necessary, why did you reintroduce it? If the method implementation returns an error, it is not an 'internal error', it's just a normal error. We can simply use the default error code of -32000 in this situation.

Comment (fjl) [reply rpc/handler.go]: Let me clarify it even more. The definition of 'internal error' is: something failed in the processing of the request. The **result** of processing can also be an error. But such errors are not considered 'internal'.

Comment (niczy) [reply rpc/handler.go]: makes sense. I reverted those changes. Thanks for reviewing!

**Pull request merged**